pilot/pkg/dns/dns.go

```diff
@@ -0,0 +1,344 @@
+		}
+
+		if len(answers) > 0 {
+			response = new(dns.Msg)
+			response.SetReply(req)
+			response.Answer = answers
+		} else {
+			response = h.queryUpstream(req)
+		}
+	}
+
+	_ = w.WriteMsg(response)
+}
+
+func (h *LocalDNSServer) Close() {
+	if h.downstreamServer != nil {
+		if err := h.downstreamServer.Shutdown(); err != nil {
+			log.Errorf("error in shutting down dns downstreamServer :%v", err)
+		}
+	}
+}
+
+func (h *LocalDNSServer) queryUpstream(req *dns.Msg) *dns.Msg {
+	var response *dns.Msg
+	for _, upstream := range h.resolvConfServers {
+		cResponse, _, err := h.upstreamClient.Exchange(req, upstream)
+		if err == nil && len(cResponse.Answer) > 0 {
+			response = cResponse
+			break
+		}
+	}
+	if response == nil {
+		response = new(dns.Msg)
+		response.SetReply(req)
+		response.Rcode = dns.RcodeNameError
+	}
+	return response
+}
+
+func separateIPtypes(ips []string) (ipv4, ipv6 []net.IP) {
+	for _, ip := range ips {
+		addr := net.ParseIP(ip)
+		if addr == nil {
+			continue
+		}
+		if addr.To4() != nil {
+			ipv4 = append(ipv4, addr.To4())
+		} else {
+			ipv6 = append(ipv6, addr)
+		}
+	}
+	return
+}
+
+func generateAltHosts(hostname string, nameinfo *nds.NameTable_NameInfo, proxyNamespace, proxyDomain string,
+	proxyDomainParts []string) []string {
+	out := make([]string, 0)
+	out = append(out, hostname+".")
+	// do not generate alt hostnames if the service is in a different domain (i.e. cluster) than the proxy
+	// as we have no way to resolve conflicts on name.namespace entries across clusters of different domains
+	if proxyDomain == "" || !strings.HasSuffix(hostname, proxyDomain) {
+		return out
+	}
+	out = append(out, nameinfo.Shortname+"."+nameinfo.Namespace+".")
+	if proxyNamespace == nameinfo.Namespace {
+		out = append(out, nameinfo.Shortname+".")
+	}
+	// Do we need to generate entries for name.namespace.svc, name.namespace.svc.cluster, etc. ?
+	// If these are not that frequently used, then not doing so here will save some space and time
+	// as some people have very long proxy domains with multiple dots
+	// For now, we will generate just one more domain (which is usually the .svc piece).
+	out = append(out, nameinfo.Shortname+"."+nameinfo.Namespace+"."+proxyDomainParts[0]+".")
+	return out
+}
+
+func (table *LookupTable) lookupHostIPv4(host string) []dns.RR {
+	ans := table.name4[host]
+	if len(ans) == 0 {
+		ans = table.cname[host]
+	}
+	return ans
+}
+
+func (table *LookupTable) lookupHostIPv6(host string) []dns.RR {
+	ans := table.name6[host]
+	if len(ans) == 0 {
+		ans = table.cname[host]
+	}
+	return ans
+}
+
+// This function stores the list of hostnames along with the precomputed DNS response for that hostname.
+// Most hostnames have a DNS response containing the A/AAAA records. In addition, this function stores a
+// variant of the host+ the first search domain in resolv.conf as the first query
+// is likely to be host.ns.svc.cluster.local (e.g., www.google.com.ns1.svc.cluster.local) due to
```

Comment:
is there any way this will end up being a bad idea? Especially if they have a custom search?

I cannot think of any but seems like an area that might cause some obscure issue in some custom environment possibly

Reply:
Oh nevermind, if coredns does this it must be a good idea